clients/src/main/java/org/apache/kafka/common/network/BrokerEndPoint.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.network;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * BrokerEndPoint is used to connect to specific host:port pair.
+ * It is typically used by clients (or brokers when connecting to other brokers)
+ * and contains no information about the security protocol used on the connection.
+ * Clients should know which security protocol to use from configuration.
+ * This allows us to keep the wire protocol with the clients unchanged where the protocol is not needed.
+ */
+public class BrokerEndPoint {
+
+    private static final Pattern URI_PARSE_EXP = Pattern.compile("\\[?([0-9a-zA-Z\\-%._:]*)]?:([0-9]+)");
```

Comment:
This seems slightly different from the original Scala version, perhaps this should be `\\[?([0-9a-zA-Z\\-%._:]*)\\]?:([0-9]+)` ?

Reply:
This is not needed, additional escape for `]` is not needed in java.

Reply:
Thanks for explanation ! I also notice that we already have a BrokerEndPointTest.scala which cover ipv6 tests... Anyway, thanks for your patience.

Reply:
Yeah, it's a good test file (BrokerEndPointTest.scala). I thought to port that too in java but it needs some zk specific scala classes to parse and uses additional scala classes which I cannot reference.